source/MRMesh/MRMeshBuilder.cpp

```diff
@@ -524,13 +528,46 @@ struct PathOverIncidentVert {
             {
                 --lastUnvisitedIndex;
                 std::iter_swap( it, vertexBegIt + lastUnvisitedIndex );
+                markVertexVisited(nextVertex);
                 return nextVertex;
             }
 
         }
         return VertId( -1 );
     }
 
+    // path = {abcDefgD} => closedPath = {DefgD}; path = {abc}
+    void extractClosedPath( std::vector<VertId>& path, std::vector<VertId>& closedPath )
+    {
+        auto lastVertex = path.back();
+        for ( size_t i = 0; i < path.size(); ++i )
+        {
+            if ( path[i] == lastVertex )
+            {
+                closedPath.reserve( path.size() - i );
+                closedPath.insert( closedPath.end(), std::make_move_iterator( path.begin() + i ),
+                                                     std::make_move_iterator( path.end() ) );
+
+                path.resize(i);
+                break;
+            }
+        }
+        for ( const auto& v : closedPath )
+            visitedVertices.reset( v );
+    }
+
+    void markVertexVisited( const VertId& v)
+    {
+        if ( visitedVertices.size() <= v )
+            visitedVertices.resize( v + 64 );
+        visitedVertices.set( v );
```

Comment:
Either resize this set in the beginning, or use autoResizeSet method

Reply:
done